### PR TITLE
Add macOS M1 runner testing configuration for PRs and scheduled tests

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
# Description

This PR configures the workflows to include tests for the new GitHub Actions M1 runners.

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
Python 3.8 and Python 3.9 are not yet available in the [version manifest](https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json), I had flagged this earlier today at https://github.com/actions/setup-python/issues/808

File: .github/workflows/test_on_push.yml [hunk 1]
```diff
@@ -180,6 +192,18 @@ jobs:
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
         python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
+        # Include macOS M1 runners
+        include:
+          # - os: macos-14
+          #   python-version: "3.8"
+          # - os: macos-14
+          #   python-version: "3.9"
```
Comment (kratman) [inline .github/workflows/test_on_push.yml:200]: Probably better to delete them than comment it out
```suggestion
```

File: .github/workflows/run_periodic_tests.yml [hunk 2]
```diff
@@ -46,6 +46,18 @@ jobs:
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
         python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
+        # Include macOS M1 runners
+        include:
+          - os: macos-14
+            python-version: "3.8"
+          - os: macos-14
+            python-version: "3.9"
```
Comment (agriyakhetarpal) [inline .github/workflows/run_periodic_tests.yml:54]:
```suggestion
```

File: .github/workflows/test_on_push.yml (hunk 1 above)
Comment (agriyakhetarpal) [reply .github/workflows/test_on_push.yml:200]: Okay sure, I don't mind either of those

File: .github/workflows/run_periodic_tests.yml [hunk 3]
```diff
@@ -175,8 +183,20 @@ jobs:
         matrix:
             os: [ubuntu-latest, macos-latest]
             python-version: ["3.8", "3.9", "3.10", "3.11"]
+          # Include macOS M1 runners
+            include:
+              - os: macos-14
+                python-version: "3.8"
+              - os: macos-14
+                python-version: "3.9"
```
Comment (agriyakhetarpal) [inline .github/workflows/run_periodic_tests.yml:191]:
```suggestion
```

File: .github/workflows/test_on_push.yml [hunk 4]
```diff
@@ -43,6 +43,18 @@ jobs:
         exclude:
           - os: ubuntu-latest
             python-version: "3.11"
+        # Include macOS M1 runners
+        include:
+          # - os: macos-14
+          #   python-version: "3.8"
+          # - os: macos-14
+          #   python-version: "3.9"
```
Comment (agriyakhetarpal) [inline .github/workflows/test_on_push.yml:51]:
```suggestion
```

**Review by kratman — COMMENTED** [review]
Looks fine as long as the new tests pass

**Comment by agriyakhetarpal** [comment]
Looks like they did, feel free to approve and merge

**Review by arjxn-py — APPROVED** [review]

**Pull request merged**